src/test/rgw/test_d4n_filter.cc

```diff
@@ -0,0 +1,2810 @@
+
+    response< int, int, 
+	     std::map<std::string, std::string>,
+	     std::map<std::string, std::string>,
+	     boost::redis::ignore_t > resp;
+
+    conn->async_exec(req, resp, yield[ec]);
+
+    ASSERT_EQ((bool)ec, false);
+    EXPECT_EQ((int)std::get<0>(resp).value(), 1);
+    EXPECT_EQ((int)std::get<1>(resp).value(), 1);
+    EXPECT_EQ(std::get<2>(resp).value().size(), 10);
+    EXPECT_EQ(std::get<3>(resp).value().size(), 10);
+
+    std::error_code err;
+    EXPECT_EQ(fs::exists(CACHE_DIR + "/" + TEST_BUCKET + testName + "/" + destName + "/test_version", err), true);  
+
+    conn->cancel();
+    testBucket->remove(env->dpp, true, optional_yield{yield});
+    DriverDestructor driver_destructor(static_cast<rgw::sal::D4NFilterDriver*>(driver));
+  }, rethrow);
+
+  io.run();
+}
+
+TEST_F(D4NFilterFixture, CopyMergeObjectRead)
+{
+  const std::string testName = "CopyMergeObjectRead";
+ 
+  boost::asio::spawn(io, [this, &testName] (boost::asio::yield_context yield) {
+    init_driver(yield);
+    create_bucket(testName, yield);
+    put_object(testName, yield);
+
+    RGWEnv rgw_env;
+    req_info info(env->cct.get(), &rgw_env);
+    rgw_zone_id zone;
+    rgw_placement_rule placement;
+    ceph::real_time mtime;
+    buffer::list bl;
+    bl.append("dest_object_version");
+    rgw::sal::Attrs attrs{{RGW_ATTR_ID_TAG, bl}};
+    std::string tag;
+
+    std::string destName = "dest_object";
+    std::unique_ptr<rgw::sal::Object> destObj = testBucket->get_object(rgw_obj_key(destName));
+    EXPECT_NE(destObj.get(), nullptr);
+
+    int ret = obj->copy_object(acl_owner,
+	     std::get<rgw_user>(owner),
+	     &info,
+	     zone,
+	     destObj.get(),
+	     testBucket.get(),
+	     testBucket.get(),
+	     placement,
+	     &mtime,
+	     &mtime,
+	     nullptr,
+	     nullptr,
+	     false,
+	     nullptr,
+	     nullptr,
+	     rgw::sal::ATTRSMOD_MERGE,
+	     false,
+	     attrs, 
+	     RGWObjCategory::Main,
+	     0,
+	     boost::none,
+	     nullptr,
+	     &tag, 
+	     &tag,
+	     nullptr,
+	     nullptr,
+	     env->dpp,
+	     optional_yield({yield}));
+    EXPECT_EQ(ret, 0);
+
+    boost::system::error_code ec;
+    request req;
+    req.push("EXISTS", TEST_BUCKET + testName + "_" + destName + "_0_0");
+    req.push("EXISTS", TEST_BUCKET + testName + "__:null_" + destName + "_0_0");
+    req.push("HGETALL", TEST_BUCKET + testName + "_" + destName + "_0_0");
+    req.push("HGETALL", TEST_BUCKET + testName + "__:null_" + destName + "_0_0");
+    req.push("FLUSHALL");
+
+    response< int, int, 
+	     std::map<std::string, std::string>,
+	     std::map<std::string, std::string>,
+	     boost::redis::ignore_t > resp;
+
+    conn->async_exec(req, resp, yield[ec]);
+
+    ASSERT_EQ((bool)ec, false);
+    EXPECT_EQ((int)std::get<0>(resp).value(), 1);
+    EXPECT_EQ((int)std::get<1>(resp).value(), 1);
+    EXPECT_EQ(std::get<2>(resp).value().size(), 10);
+    EXPECT_EQ(std::get<3>(resp).value().size(), 10);
+
```

Comment:
In all tests of CopyObj, where you test based on different flags for xattrs, you should validate whether the xattrs have been modified/not modified based on the flag passed in, like get the xattrs from the head block and see if the xattrs set are according to the expected values.

Reply:
Shouldn't this only apply to the write cache copy object tests since, in the read cache, the attributes are overwritten by the `dest_object->get_attrs()` call?